File: 206.Reverse_linked_list/reverse.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.next = None

# iterative
class Solution(object):
    def reverseList(self, head):
        """
        :type head: ListNode
        :rtype: ListNode
        """
        prev = None
        curr = head
        while curr:
            nxt = curr.next
            curr.next = prev
            prev = curr
            curr = nxt
        return prev

# recursive
class Solution(object):
    def reverseList(self, head):
        """
        :type head: ListNode
        :rtype: ListNode
        """
        if not head or not head.next:
            return head
        p = self.reverseList(head.next)
        head.next.next = head
        head.next = None
        return p

File: 206.Reverse_linked_list/reverse.go
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
package linkedlist

type ListNode struct {
	Val  int
	Next *ListNode
}

// iterative
func reverseList(head *ListNode) *ListNode {
	var prev *ListNode
	var nxt *ListNode
	curt := head
	for curt != nil {
		nxt = curt.Next
		curt.Next = prev
		prev = curt
		curt = nxt
	}
	return prev
}

//recursive
// pass the end node all the way to the beginning,
// in the meantime, reverse the link
func reverseList2(head *ListNode) *ListNode {
	if head == nil || head.Next == nil {
		return head
	}
	end := reverseList(head.Next)
	head.Next.Next = head
	head.Next = nil
	return end
}
